### DISCUSSION TITLE
Blazor Question: How to create reusable components with multiple fields?

### DISCUSSION THREAD

**Discussion by PolarbearDK** [dotnet/aspnetcore#54722] [Q&A]
I'm having trouble finding a good model for "composit components" in blazor.

Lets say we have a reusable AddressComponent (simplified)

```HTML
@inherits InputBase<AddressModel>

<InputText @bind-Value="Address.StreetAddress" />
<ValidationMessage For="@(() => Address.StreetAddress)" @bind-Value:after="@OnChanged"/>
<InputText @bind-Value="Address.Zip" />
<ValidationMessage For="@(() => Address.Zip)" @bind-Value:after="@OnChanged"/>
<InputText @bind-Value="Address.StreetAddress" />
<ValidationMessage For="@(() => Address.City" @bind-Value:after="@OnChanged"/>

@code {
    private Task OnChanged()
    {
        // Construct new AddressModel using copy-constructor to notify parent that something has changed
        var address = new AddressModel(CurrentValue);
        return ValueChanged.InvokeAsync(address);
    }
}
```
And we have a parent model (simplified - Real world has a model above this with multiple custumers, and customers has multiple addresses)
```CSHARP
public class CustomerModel
{
    public string? Email { get; set; }
    public AddressModel Address { get; set; }
}
```
And customer is shown like this:
```HTML
<EditForm ...>
<FluentValidationValidator />
<InputText @bind-Value="Customer.Email" />
<ValidationMessage For="@(() => Customer.Email)" />
<AddressComponent @bind-Value="Customer.Address" />
</EditForm>
```
Lets say that validation rules are not known to AddressComponent. It could be that address is required if no email has been provided (I'm using FluentValidation for that, and it is working well).

What is working:
- Component show model data and updates are applied to the model correctly. 
- Validation errors are shown when Submit is pressed.
- Validation errors dissapear when data is entered in a required field.

What is *not* working
- Border color doesn't change on AddressComponent InputTexts when they are in validation error state.
- Validation errors doesn't appear immediately.

Other pitfalls:
- If you use the same composit component in multiple places, the it is very easy to accidentally assign the same Id to input fields. I use a model where the component has an Id parameter that is used as a prefix for all Input Id's.

So how do you get validation to work 100% in a compound component?

**Comment by PolarbearDK** [comment]
Some comments from original poster:

[https://stackoverflow.com/questions/77491017/validation-within-nested-blazor-components](https://stackoverflow.com/questions/77491017/validation-within-nested-blazor-components) has link to [https://gist.github.com/SteveSandersonMS/090145d7511c5190f62a409752c60d00](https://gist.github.com/SteveSandersonMS/090145d7511c5190f62a409752c60d00) where @SteveSandersonMS has a possible solution to this problem.

But unfortunately, it doesnt work in this scenario (reusable component). 

The problem seems to be how `InputBase<T>` notifies EditorContext that a value has changed.
```CSHARP
EditContext?.NotifyFieldChanged(FieldIdentifier);
```
In the child component, the FieldIdentifier for the StreetAddress property becomes `Address.StreetAddress` instead of `Customer.Address.StreetAddress`. and the Validation state for the `Customer.Address.StreetAddress` is not updated correctly. 

I have however discovered that if you call EditorContext.Validate() every time a field changes, then the validation state for the nested fields are updated correctly AND bubbles down to the composite components. The downside is that you get validation errors for all untouched fields.  

I use this class as a hack for now:

```CSHARP
public class FluentValidationAllValidator : FluentValidationValidator, IDisposable
{
    [CascadingParameter] private EditContext EditContext { get; set; }

    protected override void OnInitialized()
    {
        base.OnInitialized();
        EditContext.OnFieldChanged += OnFieldChanged;
    }

    private void OnFieldChanged(object? sender, FieldChangedEventArgs e) => EditContext.Validate();

    public void Dispose() => EditContext.OnFieldChanged -= OnFieldChanged;
}
```

**Comment by SteveSandersonMS** [comment]
I don't know if it's exactly what you need, but you might want to look at inheriting from `Editor<T>` since that deals with some of the flowing of field name context. As it happens our sample of this is also an address editor: https://github.com/dotnet/aspnetcore/blob/5d31d5ed82270db90c375b78ba4f99025f0aef56/src/Components/Samples/BlazorUnitedApp/Pages/AddressEditor.razor
